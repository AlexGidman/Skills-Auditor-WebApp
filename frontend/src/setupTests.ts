// jest-dom adds custom jest matchers for asserting on DOM nodes.
// allows you to do things like:
// expect(element).toHaveTextContent(/react/i)
// learn more: https://github.com/testing-library/jest-dom
import "@testing-library/jest-dom";
import * as apiRequests from "./utility/apiRequests";
import * as reactCookie from "react-cookie";

import { MANAGER_JR, ADMIN, User } from "./utility/types";

jest.mock("uuid", () => {
    return {
        ...jest.requireActual("uuid"),
        v4: jest.fn().mockReturnValue("uuid"),
    };
});

jest.mock("react-cookie", () => {
    return {
        ...jest.requireActual("react-cookie"),
        useCookies: jest.fn(),
    };
});

export const mockApiRequests = {
    getAllCategories: jest.spyOn(apiRequests, "getAllCategories"),
    getCategory: jest.spyOn(apiRequests, "getCategory"),
    updateCategory: jest.spyOn(apiRequests, "updateCategory"),
    getUserDetails: jest.spyOn(apiRequests, "getUserDetails"),
    updateUserDetails: jest.spyOn(apiRequests, "updateUserDetails"),
    getSkill: jest.spyOn(apiRequests, "getSkill"),
    updateSkill: jest.spyOn(apiRequests, "updateSkill"),
    getAllSkills: jest.spyOn(apiRequests, "getAllSkills"),
    addUser: jest.spyOn(apiRequests, "addUser"),
    addDirectReport: jest.spyOn(apiRequests, "addDirectReport"),
    getUserFromToken: jest.spyOn(apiRequests, "getUserFromToken"),
    getAllDirectReports: jest.spyOn(apiRequests, "getAllDirectReports"),
    getAllStaffSkills: jest.spyOn(apiRequests, "getAllStaffSkills"),
    deleteUser: jest.spyOn(apiRequests, "deleteUser"),
    getStaffSkill: jest.spyOn(apiRequests, "getStaffSkill"),
    updateStaffSkill: jest.spyOn(apiRequests, "updateStaffSkill"),
    postLogin: jest.spyOn(apiRequests, "postLogin"),
    addSkill: jest.spyOn(apiRequests, "addSkill"),
    addCategory: jest.spyOn(apiRequests, "addCategory"),
    deleteStaffSkill: jest.spyOn(apiRequests, "deleteStaffSkill"),
    addUserSkill: jest.spyOn(apiRequests, "addUserSkill"),
};

export const mockAdminUser = {
    id: "1",
    firstName: "Admin",
    lastName: "User",
    email: "Admin@email.com",
    jobRole: MANAGER_JR,
    systemRole: ADMIN,
};

export const mockError = { response: { data: { error: new Error("Oops") }, status: 400 } };

beforeAll(() => {
    console.log = jest.fn();
    console.error = jest.fn();
});

beforeEach(() => {
    jest.spyOn(reactCookie, "useCookies").mockImplementation(() => [
        { token: "Bearer token" },
        jest.fn(),
        jest.fn(),
    ]);
    // @ts-ignore
    mockApiRequests.getUserFromToken.mockImplementation(() =>
        Promise.resolve({ data: mockAdminUser }),
    );
});

afterEach(() => {
    jest.resetAllMocks();
});
