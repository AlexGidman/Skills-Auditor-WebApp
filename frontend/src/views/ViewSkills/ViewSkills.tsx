import React, { useState, useEffect } from "react";
import { LoadingPlaceholder } from "../../components/LoadingPlaceholder/LoadingPlaceholder";
import { ErrorMessage } from "../../components/ErrorMessage/ErrorMessage";
import { LinkButton } from "../../components/LinkButton/LinkButton";
import { Button } from "../../components/Button/Button";
import { getSelectOptionsFromArray, useAPI, useIsAdminOrManager } from "../../utility/helper";
import { getAllSkills, deleteSkill } from "../../utility/apiRequests";
import styles from "./ViewSkills.module.css";
import { ListView } from "../../components/ListView/ListView";
import { useOutletContext, useNavigate } from "react-router-dom";
import { Skill } from "../../utility/types";
import { AppOutletContext } from "../AppWrapper/AppWrapper";

export const ViewSkills = () => {
    useIsAdminOrManager();
    const { setShowToast } = useOutletContext<AppOutletContext>();

    const navigate = useNavigate();
    const [selectedSkill, setSelectedSkill] = useState<string>();
    const [refresh, setRefresh] = useState(false);

    const { data, loading, error } = useAPI<Skill[]>(getAllSkills);
    const {
        data: deleteSkillData,
        loading: deleteSkillLoading,
        error: deleteSkillError,
        callback: executeDeleteSkill,
    } = useAPI(deleteSkill, [selectedSkill], true);

    useEffect(() => {
        if (refresh && !deleteSkillLoading) {
            setRefresh(false);
            deleteSkillError && setShowToast({ error: "Failed to delete skill" });
            deleteSkillData && navigate(0);
        }
    }, [deleteSkillLoading, deleteSkillData, deleteSkillError, navigate, refresh, setShowToast]);

    if (loading) return <LoadingPlaceholder />;
    if (error) return <ErrorMessage error={error} />;
    if (data) {
        const sortedSkills = data.sort((a, b) => a.name.localeCompare(b.name));
        return (
            <>
                <header>
                    <h1>View Skills</h1>
                </header>
                <section>
                    <ListView
                        className={styles.listview}
                        options={getSelectOptionsFromArray(
                            sortedSkills.map((skill) => {
                                return skill.id;
                            }),
                            sortedSkills.map((skill) => {
                                return `${skill.name} - ${skill.category.name}`;
                            }),
                        )}
                        value={selectedSkill}
                        labelText="Select a skill"
                        onChange={(e) => {
                            setSelectedSkill(e.target.value);
                        }}
                    />
                    <div className={styles.buttonWrapper}>
                        <LinkButton
                            className={styles.button}
                            disabled={!selectedSkill}
                            name="Edit skill"
                            path={`/edit/skill/${selectedSkill}`}
                        />
                        <Button
                            className={styles.button}
                            loading={loading || deleteSkillLoading}
                            disabled={!selectedSkill}
                            onClick={(e) => {
                                e.preventDefault();
                                executeDeleteSkill();
                                setRefresh(true);
                            }}
                        >
                            Delete Skill
                        </Button>
                    </div>
                </section>
            </>
        );
    }
    return null;
};
