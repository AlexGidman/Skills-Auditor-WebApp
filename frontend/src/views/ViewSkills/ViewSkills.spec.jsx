import { render, screen } from "@testing-library/react";
import { ViewSkills } from "./ViewSkills";
import { mockAdminUser, mockApiRequests, mockError } from "../../setupTests";
import userEvent from "@testing-library/user-event";
import * as ReactDom from "react-router-dom";
import { wait } from "@testing-library/user-event/dist/utils";

jest.mock("react-router-dom", () => {
    return {
        ...jest.requireActual("react-router-dom"),
        useOutletContext: jest.fn(),
    };
});

const mockCategory = {
    id: 1,
    name: "Category1",
};

const mockSkill1 = {
    id: 1,
    name: "Skill1",
    category: mockCategory,
};

const mockSkill2 = {
    id: 2,
    name: "Skill2",
    category: mockCategory,
};

describe("ViewSkills", () => {
    it("renders correctly when data successful", async () => {
        ReactDom.useOutletContext.mockReturnValue([mockAdminUser]);
        mockApiRequests.getAllSkills.mockImplementation(() =>
            Promise.resolve({ data: [mockSkill1, mockSkill2] }),
        );
        const { container } = render(
            <ReactDom.MemoryRouter>
                <ViewSkills />
            </ReactDom.MemoryRouter>,
        );
        expect(await screen.findByRole("heading", { name: "View Skills" })).toBeInTheDocument();

        expect(screen.getByRole("listbox")).toBeInTheDocument();
        expect(
            screen.getByRole("option", { name: `${mockSkill1.name} - ${mockCategory.name}` }),
        ).toBeInTheDocument();
        expect(
            screen.getByRole("option", { name: `${mockSkill2.name} - ${mockCategory.name}` }),
        ).toBeInTheDocument();

        expect(screen.getByRole("link", { name: "Edit skill" })).toBeInTheDocument();

        expect(container).toMatchSnapshot();
    });

    it("should have Edit skill link that has no valid href if no option is selected", async () => {
        ReactDom.useOutletContext.mockReturnValue([mockAdminUser]);
        mockApiRequests.getAllSkills.mockImplementation(() =>
            Promise.resolve({ data: [mockSkill1, mockSkill2] }),
        );
        render(
            <ReactDom.MemoryRouter>
                <ViewSkills />
            </ReactDom.MemoryRouter>,
        );
        expect(await screen.findByRole("heading", { name: "View Skills" })).toBeInTheDocument();

        const options = screen.getAllByRole("option");
        options.forEach((option) => {
            expect(option).not.toBeChecked();
        });

        expect(screen.getByRole("link", { name: "Edit skill" })).toHaveAttribute(
            "href",
            "/edit/skill/undefined",
        );
    });

    it("should have Edit skill link that has a valid href if option is selected", async () => {
        ReactDom.useOutletContext.mockReturnValue([mockAdminUser]);
        mockApiRequests.getAllSkills.mockImplementation(() =>
            Promise.resolve({ data: [mockSkill1, mockSkill2] }),
        );
        render(
            <ReactDom.MemoryRouter>
                <ViewSkills />
            </ReactDom.MemoryRouter>,
        );
        expect(await screen.findByRole("heading", { name: "View Skills" })).toBeInTheDocument();

        userEvent.selectOptions(screen.getByRole("listbox"), "1");
        await wait(100);

        expect(screen.getByRole("link", { name: "Edit skill" })).toHaveAttribute(
            "href",
            `/edit/skill/${mockSkill1.id}`,
        );
    });

    it("renders correctly when error", async () => {
        ReactDom.useOutletContext.mockReturnValue([mockAdminUser]);
        mockApiRequests.getAllSkills.mockImplementation(() => Promise.reject(mockError));
        const { container } = render(
            <ReactDom.MemoryRouter>
                <ViewSkills />
            </ReactDom.MemoryRouter>,
        );
        expect(
            await screen.findByRole("alert", { name: mockError.response.data.error.message }),
        ).toBeInTheDocument();

        expect(screen.queryByRole("listbox")).not.toBeInTheDocument();

        expect(screen.queryByRole("link", { name: "Edit skills" })).not.toBeInTheDocument();
        expect(container).toMatchSnapshot();
    });
});
