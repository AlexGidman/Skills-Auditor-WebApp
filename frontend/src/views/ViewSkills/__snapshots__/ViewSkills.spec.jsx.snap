// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ViewSkills renders correctly when data successful 1`] = `
<div>
  <header>
    <h1>
      View Skills
    </h1>
  </header>
  <section>
    <div
      class="component component listview"
    >
      <label
        class="label"
        for="Select-a skill-undefined"
      >
        Select a skill
        :
      </label>
      <select
        aria-label="Select a skill"
        class="select"
        name="Select a skill"
        size="5"
      >
        <option
          value="1"
        >
          Skill1 - Category1
        </option>
        <option
          value="2"
        >
          Skill2 - Category1
        </option>
      </select>
    </div>
    <div
      class="buttonWrapper"
    >
      <a
        class="component button disabled"
        href="/edit/skill/undefined"
      >
        Edit skill
      </a>
      <button
        class="component button"
        disabled=""
      >
        Delete Skill
      </button>
    </div>
  </section>
</div>
`;

exports[`ViewSkills renders correctly when error 1`] = `
<div>
  <div
    aria-label="Oops"
    class="component"
    role="alert"
  >
    <h1>
      Something went wrong...
    </h1>
    <p>
      Oops
    </p>
  </div>
</div>
`;
